File: python/patterns/state.py
from abc import ABC, abstractmethod


class TrafficLightState(ABC):
    @abstractmethod
    def next(self) -> "TrafficLightState":
        raise NotImplementedError

    @abstractmethod
    def color(self) -> str:
        raise NotImplementedError


class Red(TrafficLightState):
    def next(self) -> TrafficLightState:
        return Green()

    def color(self) -> str:
        return "Red"


class Green(TrafficLightState):
    def next(self) -> TrafficLightState:
        return Yellow()

    def color(self) -> str:
        return "Green"


class Yellow(TrafficLightState):
    def next(self) -> TrafficLightState:
        return Red()

    def color(self) -> str:
        return "Yellow"


class TrafficLight:
    def __init__(self) -> None:
        self.state: TrafficLightState = Red()

    def change(self) -> None:
        self.state = self.state.next()

    def show(self) -> None:
        print(f"Light is {self.state.color()}")


def run() -> None:
    print("\nState")
    light = TrafficLight()
    for _ in range(4):
        light.show()
        light.change()

File: csharp/Patterns/State.cs
using System;

namespace DesignPatterns.Patterns;

public static class StatePattern
{
    private interface ITrafficLightState
    {
        ITrafficLightState Next();
        string Name { get; }
    }

    private class Red : ITrafficLightState
    {
        public ITrafficLightState Next() => new Green();
        public string Name => "Red";
    }

    private class Green : ITrafficLightState
    {
        public ITrafficLightState Next() => new Yellow();
        public string Name => "Green";
    }

    private class Yellow : ITrafficLightState
    {
        public ITrafficLightState Next() => new Red();
        public string Name => "Yellow";
    }

    private class TrafficLight
    {
        private ITrafficLightState _state = new Red();

        public void Change() => _state = _state.Next();

        public void Show() => Console.WriteLine($"Light is {_state.Name}");
    }

    public static void Run()
    {
        Console.WriteLine("\nState");
        var light = new TrafficLight();
        for (var i = 0; i < 4; i++)
        {
            light.Show();
            light.Change();
        }
    }
}
